### ISSUE TITLE
False positive `need-mut` when mutating elements of `&mut []`

### ISSUE THREAD

**Issue by paulkre** [rust-lang/rust-analyzer#14463]
**rust-analyzer version**: v0.4.1456 (Pre-Release)
**rustc version**: rustc 1.68.0 (2c8cc3432 2023-03-06)

The following code compiles successfully but rust-analyzer displays an error:

```rust
fn func(data: &mut [f32]) {
    data[0] = 1.0; // cannot mutate immutable variable `data` | rust-analyzer(need-mut)
}
```

<img width="678" alt="Screenshot 2023-04-01 at 13 30 47" src="https://user-images.githubusercontent.com/13008492/229286233-dd994ec3-95b6-43a5-88a4-43d38d509801.png">

Marking the variable as `mut` causes `rustc` to display a warning:

<img width="848" alt="Screenshot 2023-04-01 at 13 33 57" src="https://user-images.githubusercontent.com/13008492/229286361-ae4db5c3-b129-4f05-9782-8e1672460006.png">
